File: src/main/java/me/taubsie/dungeonhub/application/enums/FlaggingApi.java
package me.taubsie.dungeonhub.application.enums;

import lombok.Getter;
import me.taubsie.dungeonhub.application.classes.FlagDetail;
import me.taubsie.dungeonhub.application.classes.FlagResponse;
import me.taubsie.dungeonhub.application.connection.FlaggingConnection;
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.function.Function;

public enum FlaggingApi {
    JERRY(
            "Jerry",
            uuid -> FlaggingConnection.getInstance().isJerryFlagged(uuid),
            aLong -> FlaggingConnection.getInstance().isJerryFlagged(aLong)
    );

    @Getter
    final String name;
    @Nullable
    final Function<UUID, Optional<FlagDetail>> uuidFunction;
    @Nullable
    final Function<Long, Optional<FlagDetail>> discordIdFunction;

    FlaggingApi(String name, @Nullable Function<UUID, Optional<FlagDetail>> uuidFunction, @Nullable Function<Long,
            Optional<FlagDetail>> discordIdFunction) {
        this.name = name;
        this.uuidFunction = uuidFunction;
        this.discordIdFunction = discordIdFunction;
    }

    public FlagResponse execute(@Nullable UUID uuid, @Nullable Long id) {
        CompletableFuture<Optional<FlagDetail>> uuidFlagged = null;
        if (uuidFunction != null && uuid != null) {
            uuidFlagged = CompletableFuture.supplyAsync(() -> uuidFunction.apply(uuid));
        }

        CompletableFuture<Optional<FlagDetail>> discordIdFlagged = null;
        if (discordIdFunction != null && id != null && id != 0) {
            discordIdFlagged = CompletableFuture.supplyAsync(() -> discordIdFunction.apply(id));
        }

        try {
            FlagDetail uuidFlaggedResponse = Optional.ofNullable(uuidFlagged)
                    .flatMap(CompletableFuture::join)
                    .orElse(null);

            FlagDetail discordIdFlaggedResponse = Optional.ofNullable(discordIdFlagged)
                    .flatMap(CompletableFuture::join)
                    .orElse(null);

            return new FlagResponse(
                    name,
                    uuidFlaggedResponse,
                    discordIdFlaggedResponse
            );
        }
        catch (CompletionException completionException) {
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Couldn't load scammer data.";
                }
            };
        }
    }
}

File: src/main/kotlin/me/taubsie/dungeonhub/application/enums/FlaggingApi.kt
package me.taubsie.dungeonhub.application.enums

import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import me.taubsie.dungeonhub.application.connection.FlaggingConnection
import me.taubsie.dungeonhub.application.exceptions.CommandExecutionException
import me.taubsie.dungeonhub.application.misc.FlagDetail
import me.taubsie.dungeonhub.application.misc.FlagResponse
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.CompletionException
import java.util.function.Function

enum class FlaggingApi(
    val displayName: String,
    val uuidFunction: Function<UUID, FlagDetail?>?,
    val discordIdFunction: Function<Long, FlagDetail?>?
) {
    JERRY(
        "Jerry",
        { uuid: UUID ->
            FlaggingConnection.isJerryFlagged(uuid)
        },
        { discordId: Long ->
            FlaggingConnection.isJerryFlagged(discordId)
        }
    ),
    HYPIXEL_SAFETY(
        "Hypixel Safety",
        { uuid: UUID ->
            FlaggingConnection.isSafetyFlagged(uuid)
        },
        { discordId: Long ->
            FlaggingConnection.isSafetyFlagged(discordId)
        }
    ),
    BLOCK_GAME(
        "Block Game Bot",
        { _ -> FlagDetail.Builder().flagged(false).build() },
        { discordId: Long ->
            FlaggingConnection.isBlockGameFlagged(discordId)
        }
    );

    private val logger: Logger = LoggerFactory.getLogger(FlaggingApi::class.java)

    fun execute(uuid: UUID?, id: Long?): FlagResponse {
        try {
            return runBlocking {
                val uuidFlagged = if (uuidFunction != null && uuid != null) {
                    async {
                        uuidFunction.apply(uuid)
                    }
                } else {
                    null
                }

                val discordIdFlagged = if (discordIdFunction != null && id != null && id != 0L) {
                    async {
                        discordIdFunction.apply(id)
                    }
                } else {
                    null
                }

                return@runBlocking FlagResponse(
                    displayName,
                    uuidFlagged?.await(),
                    discordIdFlagged?.await(),
                )
            }
        } catch (completionException: CompletionException) {
            logger.error(null, completionException)

            throw CommandExecutionException("Couldn't load scammer data.")
        }
    }
}